### DISCUSSION TITLE
FluidState Performance

### DISCUSSION THREAD

**Discussion by Dinos3396** [PaperMC/Paper#11362] [Ideas]
### Is your feature request related to a problem?

![image](https://github.com/user-attachments/assets/3fb0d3cd-318c-4262-894b-44302a71dda5)
Paper 1.21.1 57 build
Liquid spread performance is terrible, always has been. Is there nothing that can be done? That spark screenshot was captured with 200+ ticks over but I am sure you are already familiar with the performance of liquid spread.

### Describe the solution you'd like.

I dont know how liquid slope distance is calculated but it doesnt seem like its very efficient, also I am pretty sure it cannot be moved away from the main thread?

### Describe alternatives you've considered.

I tried ticking liquids less often, it works but it looks weird as it feels like it lags.

**Comment by kennytv** [comment]
Please send a full spark report, a screenshot of parts of it aren't too helpful in determining bottlenecks

**Reply by Dinos3396** [reply]
> Please send a full spark report, a screenshot of parts of it aren't too helpful in determining bottlenecks

https://spark.lucko.me/uMyNwpyxyg

**Comment by Leguan16** [comment]
Paper generally does not provide support cracked servers. (Offline Mode + Skins restorer)

**Reply by Dinos3396** [reply]
> Paper generally does not provide support cracked servers. (Offline Mode + Skins restorer)

I didnt ask for support, this is a discussion. I am interested in the topic so if any developers feel like it, I would love to hear their thoughts and insights about the topic.

**Comment by Dinos3396** [comment]
Would something like this help or cause more issues?

```
protected int getSlopeDistance(LevelReader world, BlockPos pos, int initialDistance, Direction excludedDirection, BlockState state, BlockPos fromPos) {
    Queue<BlockPos> queue = new LinkedList<>();
    Map<BlockPos, Integer> distanceMap = new HashMap<>();
    queue.add(pos);
    distanceMap.put(pos, 0);

    while (!queue.isEmpty()) {
        BlockPos currentPos = queue.poll();
        int currentDistance = distanceMap.get(currentPos);

        if (currentDistance >= this.getSlopeFindDistance(world)) {
            continue;
        }

        for (Direction direction : Plane.HORIZONTAL) {
            if (direction == excludedDirection) continue;

            BlockPos nextPos = currentPos.relative(direction);
            BlockState nextState = world.getBlockState(nextPos);
            FluidState nextFluidState = nextState.getFluidState();

            if (this.canPassThrough(world, this.getFlowing(), currentPos, state, direction, nextPos, nextState, nextFluidState)) {
                boolean canFlowDown = this.isWaterHole(world, this.getFlowing(), nextPos, nextState, nextPos.below(), world.getBlockState(nextPos.below()));

                if (canFlowDown) {
                    return currentDistance + 1;
                }

                if (!distanceMap.containsKey(nextPos)) {
                    queue.add(nextPos);
                    distanceMap.put(nextPos, currentDistance + 1);
                }
            }
        }
    }

    return 1000;
}

private void spreadToSides(Level world, BlockPos pos, FluidState fluidState, BlockState blockState) {
    int amount = fluidState.getAmount() - this.getDropOff(world);
    if ((Boolean) fluidState.getValue(FALLING)) {
        amount = 7;
    }

    if (amount > 0) {
        for (Direction direction : Plane.HORIZONTAL) {
            BlockPos nextPos = pos.relative(direction);
            BlockState nextBlockState = world.getBlockState(nextPos);
            FluidState nextFluidState = world.getFluidState(nextPos);

            if (this.canSpreadTo(world, pos, blockState, direction, nextPos, nextBlockState, nextFluidState, this.getFlowing())) {
                this.spreadTo(world, nextPos, nextBlockState, direction, this.getFlowing(amount, false));
            }
        }
    }
}
```

So it no longer uses recursion and uses a loop instead and it doesnt check the same blocks over and over again.

**Comment by Malfrador** [comment]
Is your world pre-generated? I've only ever seen fluid ticking being that much of an issue during world generation. And thats easily fixed by pre-generating the world. 

I don't think its worth making large changes here, as this really seems to be an edge case. Fluids taking more time than entities or chunks in a spark report is something I've never seen before in years of looking at performance reports.
